Client/DeviceCheck/Cryptography.swift

```diff
@@ -0,0 +1,240 @@
+    -----END PUBLIC KEY-----
+    """
+    return result
+  }
+  
+  /// Deletes the key from the secure-enclave and keychain
+  @discardableResult
+  public func delete() -> Error? {
+    let error = SecItemDelete([
+      kSecClass: kSecClassKey,
+      kSecAttrApplicationTag: keyId.data(using: .utf8)!
+    ] as CFDictionary)
+    
+    if error == errSecSuccess || error == errSecItemNotFound {
+      return nil
+    }
+    
+    return CryptographyError(code: error)
+  }
+  
+  /// Signs a "message" with the key and returns the signature
+  public func sign(message: String) throws -> Data {
+    var error: Unmanaged<CFError>?
+    let signature = SecKeyCreateSignature(key,
+                                          .ecdsaSignatureMessageX962SHA256,
+                                          message.data(using: .utf8)! as CFData,
+                                          &error)
+    
+    if let error = error?.takeUnretainedValue() {
+      throw error as Error
+    }
+    
+    guard let result = signature as Data? else {
+      throw CryptographyError(code: -1, description: "Cannot sign message with cryptographic key.")
+    }
+    
+    return result
+  }
+}
+
+/// A class used for generating cryptographic keys
+public class Cryptography {
+  
+  /// The access control flags for any keys generated
+  public static let accessControlFlags = SecAccessControlCreateWithFlags(kCFAllocatorDefault, kSecAttrAccessibleWhenUnlockedThisDeviceOnly, [.privateKeyUsage], nil) //.biometryAny
+  
+  /// Determines if a key exists in the keychain without retrieving it
+  public class func keyExists(id: String) -> Bool {
+    let query: [CFString: Any] = [
+      kSecClass: kSecClassKey,
+      kSecAttrApplicationTag: id.data(using: .utf8)!,
+      kSecMatchLimit: kSecMatchLimitOne,
+      kSecReturnRef: kCFBooleanFalse as Any,
+      kSecReturnAttributes: kCFBooleanTrue as Any
+    ]
+    
+    var result: CFTypeRef?
+    let error = SecItemCopyMatching(query as CFDictionary, &result)
+    return error == errSecSuccess || error == errSecInteractionNotAllowed
+      && result != nil && CFDictionaryGetCount((result! as! CFDictionary)) > 0 //swiftlint:disable:this force_cast
+  }
+  
+  /// Determines if a key requires biometrics to access
+  public class func isKeyRequiringBiometrics(id: String) -> Bool {
+    let query: [CFString: Any] = [
+      kSecClass: kSecClassKey,
+      kSecAttrApplicationTag: id.data(using: .utf8)!,
+      kSecMatchLimit: kSecMatchLimitOne,
+      kSecReturnRef: kCFBooleanFalse as Any,
+      kSecReturnAttributes: kCFBooleanTrue as Any
+    ]
+    
+    var result: CFTypeRef?
+    let error = SecItemCopyMatching(query as CFDictionary, &result)
+    if error == errSecSuccess || error == errSecInteractionNotAllowed {
+      if let result = result as? [String: Any], let accessControl = result[kSecAttrAccessControl as String] as! SecAccessControl? { //swiftlint:disable:this force_cast
```

Comment:
![image](https://user-images.githubusercontent.com/1530031/67502936-bf60f900-f654-11e9-8d30-47718b0754d2.png)
<img width="742" alt="Screen Shot 2019-10-24 at 11 51 58 AM" src="https://user-images.githubusercontent.com/1530031/67502939-c25be980-f654-11e9-9d38-fa9766ca1287.png">

This is a bug in Xcode and it can't be done any other way. I've tried to `if let`, I have tried `guard let`, I have tried optional cast and ternary unwrap.. None of them work and will give the above errors. That's why I have to force cast it to an optional type and then unwrap that optional type :(

Reply:
Lol XCode really doesn't like you :D
Would be good to put a comment about it somewhere in code

Reply:
I fixed it by first checking what type was returned via its TypeId and then force-casting. This should provide the safety we're looking for.